File: opensource/data/src/main/java/com/sos/scheduler/engine/data/order/OrderKey.java
package com.sos.scheduler.engine.data.order;

import com.google.common.base.Objects;
import com.sos.scheduler.engine.data.folder.JobChainPath;
import org.codehaus.jackson.annotate.JsonProperty;

import static com.google.common.base.Objects.equal;

public class OrderKey {
    private final JobChainPath jobChainPath;
    private final OrderId id;

    public OrderKey(@JsonProperty("jobChainPath") JobChainPath jobChainPath, @JsonProperty("id") OrderId id) {
        this.id = id;
        this.jobChainPath = jobChainPath;
    }

    public final JobChainPath getJobChainPath() {
        return jobChainPath;
    }

    public final String jobChainPathString() {
        return jobChainPath.asString();
    }

    public final OrderId getId() {
        return id;
    }

    public final String idString() {
        return id.asString();
    }

    @Override public int hashCode() {
        return Objects.hashCode(jobChainPath, id);
    }

    @Override public boolean equals(Object o) {
        return o instanceof OrderKey && eq((OrderKey)o);
    }

    private boolean eq(OrderKey o) {
        return equal(jobChainPath, o.jobChainPath) && equal(id, o.id);
    }

    @Override public String toString() {
        return jobChainPath+":"+id;
    }

    public static OrderKey of(String jobChainPath, String id) {
        return new OrderKey(JobChainPath.of(jobChainPath), new OrderId(id));
    }
}

File: opensource/data/src/main/scala/com/sos/scheduler/engine/data/order/OrderKey.scala
package com.sos.scheduler.engine.data.order

import OrderKey._
import com.fasterxml.jackson.annotation.{JsonProperty, JsonCreator}
import com.sos.scheduler.engine.data.folder.{FileBasedType, TypedPath, JobChainPath}

// Wegen "scala: error while loading IsString, class file OrderKey.class":  @JsonSerialize(using = classOf[OrderKeySerializer])
final case class OrderKey(jobChainPath: JobChainPath, id: OrderId)
extends SerializableOrderKey
with TypedPath {

  def typ =
    FileBasedType.order

  def string =
    jobChainPath.string + separator + id

  @Deprecated
  def getJobChainPath: JobChainPath =
    jobChainPath

  @Deprecated
  def jobChainPathString: String =
    jobChainPath.string

  @Deprecated
  def getId: OrderId =
    id

  @Deprecated
  def idString: String =
    id.string

  override def toString =
    s"${jobChainPath.string}:$id"
}


object OrderKey {
  private val separator = ','

  def apply(o: String): OrderKey = {
    val i = o indexOf ','
    require(i > 0, "OrderKey TypedPath needs comma ',' to separate job chain path from order key")
    apply(o.substring(0, i), o.substring(i + 1))
  }

  @JsonCreator
  def apply(
      @JsonProperty("jobChainPath") jobChainPath: String,
      @JsonProperty("id") id: String): OrderKey =
    OrderKey(JobChainPath(jobChainPath), OrderId(id))

  def of(jobChainPath: String, id: String): OrderKey =
    OrderKey(JobChainPath(jobChainPath), OrderId(id))
}
